src/Chainweb/Pact/PactService.hs

```diff
@@ -477,97 +480,135 @@ finalizeCheckpointer finalize = do
+attemptBuyGas cp miner txs = withDiscardedBatch $ do
+  mbLatestBlock <- liftIO $ _cpGetLatestBlock cp
+  (bh, bhash) <- case mbLatestBlock of
+                     Nothing -> throwM NoBlockValidatedYet
+                     (Just !p) -> return p
+  let target = Just (succ bh, bhash)
+  withCheckpointer target "validateChainwebTxs/attemptBuyGas" $ \(PactDbEnv' dbEnv) -> do
+      psEnv <- ask
+      res <- V.fromList . ($ []) . sfst <$> V.foldM (f psEnv dbEnv) (T2 id mempty) txs
+      return $! Discard res
+  
+  where
+    f psEnv dbEnv (T2 dl mcache) cmd = do
+      T2 !res mcache' <- runBuyGas psEnv dbEnv mcache cmd
+      pure $! T2 (dl . (res :)) mcache'
+
+    createGasEnv
+        :: PayloadCas cas
+        => PactServiceEnv cas
+        -> P.PactDbEnv a
+        -> P.Command (P.Payload P.PublicMeta P.ParsedCode)
+        -> P.CommandEnv a
+    createGasEnv envM db cmd =
+      P.CommandEnv Nothing P.Transactional db logger
+      freeGasEnv publicData spv nid
+      where
+        !logger = _cpeLogger . _psCheckpointEnv $ envM
+        !publicData = set P.pdPublicMeta (publicMetaOf cmd) (_psPublicData envM)
+        !spv = _psSpvSupport envM
+        !nid = networkIdOf cmd
+
+    runBuyGas
+        :: PayloadCas cas
+        => PactServiceEnv cas
+        -> P.PactDbEnv a
+        -> ModuleCache
+        -> (ChainwebTransaction, Uniqueness)
+        -> PactServiceM cas (T2 (ChainwebTransaction, Uniqueness, Bool) ModuleCache)
+    runBuyGas _ _ mcache (tx, u@Duplicate) = return $ T2 (tx, u, False) mcache
+    runBuyGas envM db mcache (tx, u@Unique) = do
+      let cmd = _payloadObj <$> tx
+          gasPrice = gasPriceOf cmd
+          gasLimit = fromIntegral $ gasLimitOf cmd
+          supply = gasFeeOf gasLimit gasPrice
+          buyGasEnv = createGasEnv envM db cmd 
+
+      buyGasResultE <- liftIO $! P.catchesPactError $!
+        buyGas buyGasEnv cmd miner supply mcache
+
+      case buyGasResultE of
+        Left _ -> return $ T2 (tx, u, False) mcache
+        Right (Left _) -> return $ T2 (tx, u, False) mcache
+        Right (Right (T3 _ _ mcache')) -> return $ T2 (tx, u, True) mcache'
+
 -- | The principal validation logic for groups of Pact Transactions.
 --
 -- Skips validation for genesis transactions, since gas accounts, etc. don't
 -- exist yet.
 --
 validateChainwebTxs
-    :: PactDbEnv'
+    :: PayloadCas cas
+    => PactServiceEnv cas
+    -> PactServiceState
     -> Checkpointer
+    -> Miner
     -> BlockCreationTime
     -> BlockHeight
     -> Vector ChainwebTransaction
     -> IO (Vector Bool)
-validateChainwebTxs dbEnv cp blockOriginationTime bh txs
+validateChainwebTxs psEnv psState cp miner blockOriginationTime bh txs
     | bh == 0 = pure $! V.replicate (V.length txs) True
     | V.null txs = pure V.empty
     | otherwise = do
           let f t = let p = view P.cmdHash t
                     in (bool Unique Duplicate . isJust) <$> _cpLookupProcessedTx cp p
```

Comment:
Suggesting

```haskell
let f t = (bool Unique Duplicate . isJust) <$!> _cpLookupProcessedTx cp (P._cmdHash t)
```

Instead to avoid a thunk and strict map to execute immediately